### ISSUE TITLE
Missing part for file causes filling with zero to insane size

### ISSUE THREAD

**Issue by Longoon12000** [Lazza/RecuperaBit#75]
Hello

I am currently restoring all (non-deleted) files from a corrupted NTFS partition. At some point it has reached a file with the following message:

```WARNING:root:Missing part for File(#1544, ^^1516^^, DCIM000105.avi, offset = 6294607 sectors), filling 844632251189574 clusters with zeros```

The original file is less than 1GB in size but while I was gone it has written over 900GB in this file and hasn't stopped since. I've set a `watch -n 10` to truncate the file to 0 bytes every 10 seconds but I fear that the amount it is writing is so big that I will have to wait until the heat death of the universe or my hard drive that is being written to gives out.

Is there anything I can do about this?

**Comment by Lazza** [comment]
> The original file is less than 1GB in size

Unfortunately, the record on disk seems to contain a different value from the size. It is probably corrupted but there is no reliable way to detect whether the reported size is correct or not.

You might try to monkey-patch the code to skip the filling operation if the number of clusters is above a certain threshold.

**Comment by Longoon12000** [comment]
Thanks for the quick reply!

I've adjusted /fs/ntfs.py and added a line at line 335:

```
            if diff > 0:
                logging.warning(
                    u'Missing part for {}, filling {} clusters '
                    'with zeros'.format(self, diff)
                )
                diff = 1
```

generally setting the diff to 1 so files that would require filling are only filled with one zero. the file is corrupt anyways so I don't need it filled up to the original size.

Does that make sense? I've tested it on just that file and a working one and it seems to have had the desired effect while not affecting the working file, but since I had to restart the restore process I'll have to wait about 8 hours until reaching this file again to verify that I didn't break anything else.

**Comment by Lazza** [comment]
> Does that make sense?

I might consider adding an option to skip the zerofill altogether or even check if the missing size is "too big" for some kind of value.

> I'd like to say that this program is great and doing better than the 60$ version of wondershare recoverit

Thank you for the kind words. 😄 I have to admit that the textual interface is cumbersome and I should be giving much more attention to the program than what I am able to give now (due to lack of time). I am planning to rewrite the structure of the save file by using a SQLite database and the TUI should get revamped as well.

On this regard, I have submitted a presentation proposal for OSDF2020, in case you want to vote for it. 😉 
https://www.surveymonkey.com/r/osdfconvoting2020

**Comment by Longoon12000** [comment]
Firstly, sorry for the wall of text! I'm going a little bit off topic :)

> I might consider adding an option to skip the zerofill altogether or even check if the missing size is "too big" for some kind of value.

Too big is very subjective and can vary widely depending on the file types, eg. restoring a drive full of database files "too big" can be in the order of 100GB+ while a folder of images on the same drive would consider 100MB "too big" for a file, these don't really mix. That's a tough one to solve.

I'm not much of a recovery guy myself so I don't know all the applications of recovery, for me skipping the zero fill was completely sufficient since the file was corrupted anyways and I wouldn't have known what to do with the file filled up with zeroes.

The code that I added only worked partially and caused some other issues or didn't completely solve it (the content_iterator function, if I recall correctly, also required adjustment) and I had to restart a bunch, but adding the check to skip files that already existed in the output folder helped with that. Here I also very much appreciate that your tool stores the results of the analysis and didn't have to re-analyse every time I restarted it (like some other worthless 60$ application does...).

All in all I was able to recover 868GB and only 26 files failed to recover, so again thank you so much!

> I have to admit that the textual interface is cumbersome

Honestly I don't mind the console interface, it's well documented and usable!

> On this regard, I have submitted a presentation proposal for OSDF2020, in case you want to vote for it. 😉
> https://www.surveymonkey.com/r/osdfconvoting2020

I voted and really hope the proposal gets accepted!
